### ISSUE TITLE
Asset issues when upgrading to Joomla 4

### ISSUE THREAD

**Issue by wilsonge** [joomla/joomla-cms#41359]
### Steps to reproduce the issue
Upgrade a site from Joomla 3 to Joomla 4. After the upgrade you'll notice several items missing required assets

- [x] Components (should be fixed with https://github.com/joomla/joomla-cms/pull/41358)
- [ ] Modules (mainly the new administrator module included - guided tours, frontend link, dashboard modules and more - we'll need to work through the upgrade scripts for a complete list)
- [ ] Workflow assets (all core transitions and stages)

### Expected result
No asset issues on upgrade

### Actual result
Missing assets on upgrade

### Additional comments
Issues found thanks to the [PWT ACL Manager](https://extensions.joomla.org/extension/acl-manager/) whilst upgrading our core sites

**Comment by richard67** [comment]
@wilsonge We don’t insert new assets with update SQL scripts because we can’t really know the right lft and rgt values when inserting them.

**Comment by wilsonge** [comment]
I know but that's why we've had the refresh assets in script.php since joomla update was introduced in 2.5

**Comment by chmst** [comment]
After updating a 4.3.x to the news stable version, all backend modules with assed_id = 0 were duplicated. I updated with upload and update. 
![grafik](https://github.com/joomla/joomla-cms/assets/1035262/5560519f-3dc2-4b00-a13c-daf399b6e9ce)

Table Before 
![grafik](https://github.com/joomla/joomla-cms/assets/1035262/df632e76-ab7f-4fe1-97b6-30e32fde9ba4)
Table After
![grafik](https://github.com/joomla/joomla-cms/assets/1035262/1043d75d-f4b4-40c2-ae30-1cd26f74d5ca)
![grafik](https://github.com/joomla/joomla-cms/assets/1035262/5560519f-3dc2-4b00-a13c-daf399b6e9ce)

**Comment by richard67** [comment]
> After updating a 4.3.x to the news stable version, all backend modules with assed_id = 0 were duplicated. I updated with upload and update. This will result in tons of user issues and so I think it needs a quick solution.

I don't think this is related to the asset_ids, and I never have seen that with a clean install. Maybe there was something wrong with the schema version in database so the SQL update script has run a 2nd time?

**Comment by chmst** [comment]
I never had it on fresh installation, of course. 
And not on installations which were set up with J4. This is an old site, upgraded from J3 in many steps. 
The schema version could be a hint, thank you.

**Comment by chmst** [comment]
I found the issue when I updated  the site (local copy). Not testing any PR. 
I was wondering how my backend looks after the update and  noticed is that only modules with asset_id = 0 are duplicated

**Comment by richard67** [comment]
@chmst The modules are inserted with this update script here: https://github.com/joomla/joomla-cms/blob/4.3-dev/administrator/components/com_admin/sql/updates/mysql/4.0.0-2019-07-13.sql#L29-L107

So if they have been inserted twice, the schema version in database was either missing or smaller than "4.0.0-2019-07-13".

The schema version of the core is the record in table `#__schemas` which belongs to the extension named 'files_joomla'.

**Comment by richard67** [comment]
@chmst Do you have SQL update scripts with the following names in folder "administrator/components/com_admin/sql/updates/mysql":
- 4.0.0-2019-01-29.sql
- 4.0.0-2019-02-17.sql
- 4.0.0-2019-02-20.sql
- 4.0.0-2019-05-07.sql

If these 4 scripts exists, then they are the problem. They have been combined together into a new script "4.0.0-2019-07-13.sql" with PR #25570 before 4.0.0 stable, i.e. during alpha or beta phase. But this would mean that the duplicate modules have been already created when updating to 4.0, and that it was an alpha or beta which was used for that.

Or it is like I wrote before that the schema version was not ok.

**Comment by richard67** [comment]
@chmst Another way how the duplicate modules can happen is when you have a failed update and you restore a backup from before the update, but you haven't used an empty database for that, and then you update again, this time successful.

**Comment by chmst** [comment]
Thanks for investigating. I don't have these files. I will repeat the udate on a new copy of my site and see what happens.

**Comment by brianteeman** [comment]
FYI I have seen a few reports recently the same as @chmst

**Comment by richard67** [comment]
> FYI I have seen a few reports recently the same as @chmst

Me too, but as I've never had a pre-update backup to check that, I don't really know the reason for the issue. I only can make assumption based on my knowledge on how it should work and where problems could come from.
